### PR TITLE
Resolvev1destination

### LINKED ISSUE #890: URIResolver uses v1beta1.Destination instead of v1
/area API
/kind cleanup

## Expected Behavior

URIResolver uses v1.Destination

## Actual Behavior

URIResolver uses v1beta1.Destination, requiring code using it to translate between the two types.

## Additional Info
Alternately, if v1beta1.Destination is the right type for URIResolver, add method(s) to Destination types that translate between them.

### PR DESCRIPTION
**Pull request by vaikas**
Fixes #890 

Add a new method URIFromDestinationV1 for resolving v1.Destination
Since v1beta1 and v1 Addressable are the same, watch for v1 duck addressable.

### CODE REVIEW

File: resolver/addressable_resolver.go [hunk 1]
```diff
@@ -103,6 +113,33 @@ func (r *URIResolver) URIFromDestination(dest duckv1beta1.Destination, parent in
...
 
+// URIFromDestinationV1 resolves a v1.Destination into a URI string.
+func (r *URIResolver) URIFromDestinationV1(dest duckv1.Destination, parent interface{}) (string, error) {
+	if dest.Ref != nil {
+		url, err := r.URIFromObjectReference(dest.Ref, parent)
+		if err != nil {
+			return "", err
+		}
+		if dest.URI != nil {
+			if dest.URI.URL().IsAbs() {
```
Comment (n3wscott) [inline resolver/addressable_resolver.go]: I thought we loosened this to just use `ResolveReference` when not nil and if the uri is a full rewrite, so be it.

Comment (vaikas) [reply resolver/addressable_resolver.go]: The old one still uses it above, so I didn't think so? I'd rather decouple that change from this though if that's cool?

File: resolver/addressable_resolver.go [hunk 2]
```diff
@@ -103,6 +105,33 @@ func (r *URIResolver) URIFromDestination(dest duckv1beta1.Destination, parent in
...
+				return "", fmt.Errorf("absolute URI is not allowed when Ref or [apiVersion, kind, name] exists")
```
Comment (vagababov) [inline resolver/addressable_resolver.go]: nit: no formatting, so just `return "", errors.New("...")`.

File: resolver/addressable_resolver.go [hunk 3]
```diff
@@ -103,6 +105,33 @@ func (r *URIResolver) URIFromDestination(dest duckv1beta1.Destination, parent in
...
+			}
+			return url.URL().ResolveReference(dest.URI.URL()).String(), nil
+		}
+		return url.URL().String(), nil
+	}
+
+	if dest.URI != nil {
+		// IsAbs check whether the URL has a non-empty scheme. Besides the non non-empty scheme, we also require dest.URI has a non-empty host
+		if !dest.URI.URL().IsAbs() || dest.URI.Host == "" {
+			return "", fmt.Errorf("URI is not absolute(both scheme and host should be non-empty): %v", dest.URI.String())
```
Comment (vagababov) [inline resolver/addressable_resolver.go]:
```suggestion
			return "", fmt.Errorf("URI is not absolute(both scheme and host should be non-empty): %q", dest.URI.String())
```

File: resolver/addressable_resolver.go [hunk 4]
```diff
@@ -103,6 +105,33 @@ func (r *URIResolver) URIFromDestination(dest duckv1beta1.Destination, parent in
...
+		}
+		return dest.URI.String(), nil
+	}
+
+	return "", fmt.Errorf("destination missing Ref and URI, expected at least one")
```
Comment (vagababov) [inline resolver/addressable_resolver.go]: ditto here re: `errors.New`

File: resolver/addressable_resolver_test.go [hunk 5]
```diff
@@ -360,6 +361,233 @@ func TestGetURI_ObjectReference(t *testing.T) {
...
+			r := resolver.NewURIResolver(ctx, func(types.NamespacedName) {})
+
+			// Run it twice since this should be idempotent. URI Resolver should
+			// not modify the cache's copy.
+			_, _ = r.URIFromDestinationV1(tc.dest, getAddressable())
+			uri, gotErr := r.URIFromDestinationV1(tc.dest, getAddressable())
+
+			if gotErr != nil {
+				if tc.wantErr != nil {
+					if diff := cmp.Diff(tc.wantErr.Error(), gotErr.Error()); diff != "" {
```
Comment (vagababov) [inline resolver/addressable_resolver_test.go]: `.Error()` is a string, just compare with `got != want`.

File: resolver/addressable_resolver_test.go [hunk 6]
```diff
@@ -360,6 +361,233 @@ func TestGetURI_ObjectReference(t *testing.T) {
...
+						t.Errorf("%s: unexpected error (-want, +got) = %v", n, diff)
+					}
+				} else {
+					t.Errorf("%s: unexpected error: %v", n, gotErr.Error())
```
Comment (vagababov) [inline resolver/addressable_resolver_test.go]: `t.Run` already specifies the test name; `%s: ` in the beginning is redundant.

File: resolver/addressable_resolver_test.go [hunk 7]
```diff
@@ -360,6 +361,233 @@ func TestGetURI_ObjectReference(t *testing.T) {
...
+				}
+			}
+			if gotErr == nil {
```
Comment (vagababov) [inline resolver/addressable_resolver_test.go]: do a continue in the `if gotErr != nil` above, then remove if here and unindent.

File: resolver/addressable_resolver_test.go [hunk 8]
```diff
@@ -360,6 +361,233 @@ func TestGetURI_ObjectReference(t *testing.T) {
...
+				got := uri
+				if diff := cmp.Diff(tc.wantURI, got); diff != "" {
```
Comment (vagababov) [inline resolver/addressable_resolver_test.go]: same here: strings just `!=` and no need to print the test name.

File: resolver/addressable_resolver_test.go [hunk 9]
```diff
@@ -360,6 +361,233 @@ func TestGetURI_ObjectReference(t *testing.T) {
 	}
 }
 
+func TestGetURI_DestinationV1(t *testing.T) {
+	tests := map[string]struct {
+		objects []runtime.Object
+		dest    duckv1.Destination
+		wantURI string
+		wantErr error
```
Comment (vagababov) [inline resolver/addressable_resolver_test.go]: since you're comparing error strings, just make `wantErr` a string.

**Comment by n3wscott** [comment]
/lgtm
/hold
hold for if you want to follow up on the victor comments 👍

File: resolver/addressable_resolver_test.go (hunk 5 above)
Comment (vaikas) [reply resolver/addressable_resolver_test.go]: Sorry, are you saying get rid of cmp.Diff? I like the fact the diff actually makes it easier to understand what the diffs are and not just that they are different. Or maybe I misunderstood what you meant here :)

**Comment by vaikas** [comment]
Thanks, @vagababov PTAL.

File: resolver/addressable_resolver_test.go [hunk 10]
```diff
@@ -58,14 +59,14 @@ func init() {
...
+func TestGetURI_DestinationV1Beta1(t *testing.T) {
 	tests := map[string]struct {
 		objects []runtime.Object
 		dest    duckv1beta1.Destination
 		wantURI string
-		wantErr error
+		wantErr string
 	}{"nil everything": {
-		wantErr: fmt.Errorf("destination missing Ref, [apiVersion, kind, name] and URI, expected at least one"),
+		wantErr: fmt.Sprintf("destination missing Ref, [apiVersion, kind, name] and URI, expected at least one"),
```
Comment (vagababov) [inline resolver/addressable_resolver_test.go]:
```suggestion
		wantErr: "destination missing Ref, [apiVersion, kind, name] and URI, expected at least one",
```

File: resolver/addressable_resolver_test.go [hunk 11]
```diff
@@ -100,7 +101,7 @@ func TestGetURI_ObjectReference(t *testing.T) {
 				DeprecatedAPIVersion: addressableAPIVersion,
 				DeprecatedNamespace:  testNS,
 			},
-			wantErr: fmt.Errorf("ref and [apiVersion, kind, name] can't be both present"),
+			wantErr: fmt.Sprintf("ref and [apiVersion, kind, name] can't be both present"),
```
Comment (vagababov) [inline resolver/addressable_resolver_test.go]:
```suggestion
			wantErr: "ref and [apiVersion, kind, name] can't be both present",
```
And same below where there's no formatting.

File: resolver/addressable_resolver_test.go [hunk 12]
```diff
@@ -298,36 +299,36 @@ func TestGetURI_ObjectReference(t *testing.T) {
...
 			objects: []runtime.Object{
 				getAddressableNoStatus(),
 			},
 			dest:    duckv1beta1.Destination{Ref: getUnaddressableRef()},
-			wantErr: fmt.Errorf(`address not set for %+v`, getUnaddressableRef()),
+			wantErr: fmt.Sprintf(`address not set for %+v`, getUnaddressableRef()),
 		}, "notFound": {
 			dest:    duckv1beta1.Destination{Ref: getUnaddressableRef()},
-			wantErr: fmt.Errorf(`failed to get ref %+v: %s "%s" not found`, getUnaddressableRef(), unaddressableResource, unaddressableName),
+			wantErr: fmt.Sprintf(`failed to get ref %+v: %s "%s" not found`, getUnaddressableRef(), unaddressableResource, unaddressableName),
```
Comment (vagababov) [inline resolver/addressable_resolver_test.go]: here instead of `"%s"`   you can just do `%q`.

**Pull request merged**